### DISCUSSION TITLE
Packaging libpardiso

### DISCUSSION THREAD

**Discussion by ivan-pi** [fortran-lang/fpm#460] [General]
[PARDISO](https://www.pardiso-project.org/) is a fast solver for sparse systems of linear equations. Academic users can receive a free license valid for a period 12-months with some special conditions.

Upon registering, the PARDISO homepage allows you to downloaded two shared library objects for x86-64, 64-bit Linux operating system:
* `libpardiso720-GNU831-X86-64.so`
* `libpardiso720-INTEL120-X86-64.so`

No modules are given, but a [reference sheet](https://pardiso-project.org/manual/refsheet.pdf) for the interface is available, and also a longer [user guide](https://pardiso-project.org/manual/manual.pdf) with instructions about linking.

Upon downloading the package, each licensee recieves a file `pardiso.lic`, which

> must always be present either in the home directory of the user or in the directory from which the user runs a program linked with the PARDISO library. Alternatively, this file can be placed in a fixed location and the environment variable PARDISO_LIC_PATH must be set to the path of its location.

Since PARDISO is provided as a shared library, the environment variable LD_LIBRARY_PATH must point to the PARDISO library.  

Following the PARDISO user guide, I've created the attached interface module ([pardiso_interface.f90.txt](https://github.com/fortran-lang/fpm/files/6370073/pardiso_interface.f90.txt); delete the `.txt` suffix).

My current set up has been to place the shared library and license file in the folder `/opt/pardiso`, and add the following commands to my `~/.bashrc` file:
```
# PARDISO SOLVER
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/pardiso
export PARDISO_LIC_PATH=/opt/pardiso
# Suppress PARDISO License message
export PARDISOLICMESSAGE=1
```
Afterwards I will just hardwire the shared libraries into my (C)Make files, for each compiler separately.

What is the easiest way to expose PARDISO as a (shareable) `fpm` project, given that each user needs to download his own copies and license file?

**Comment by awvwgk** [comment]
Intel is shipping a version of libpardiso in MKL and is also providing interfaces as Fortran include files:
```
❯ fd pardiso /opt/intel
/opt/intel/oneapi/mkl/2021.1.1/include/mkl_pardiso.f90
/opt/intel/oneapi/mkl/2021.1.1/include/mkl_pardiso.fi
/opt/intel/oneapi/mkl/2021.1.1/include/mkl_pardiso.h
```

There is also [`pardiso4py`](https://github.com/ma-sadeghi/pardiso4py), looking at [its conda-forge feedstock](https://github.com/conda-forge/pardiso4py-feedstock/blob/master/recipe/meta.yaml) it is apparently using MKL as libpardiso provider.

**Reply by ivan-pi** [reply]
I believe the MKL Pardiso is an older fork of the project and at least for some problems much slower looking at the benchmarks on the PARDISO homepage. The website says:

> Important: Please note that the Intel MKL version of PARDISO is based on our version from 2006 and that a lot of new features and improvements of PARDISO are not available in the Intel MKL library.

I believe the interfaces and settings have also diverged slightly, so it might be best to consider them as two different libraries at this point.

**Comment by awvwgk** [comment]
Well, reading the license personal use is the only thing that is really covered, installing libpardiso is therefore left as exercise to the user.

One viable strategy here might be dlopening libpardiso rather than linking against it, this gives you the possibility to react on a missing libpardiso installation and maybe fall back to a different backend or print a more helpful error message than ld.

**Reply by ivan-pi** [reply]
Interesting idea! Rosetta code has a rather complete example of using `dlopen` from Fortran: https://rosettacode.org/wiki/Call_a_function_in_a_shared_library#Fortran

Presumably, I could then use the output of [`compiler_version`](https://gcc.gnu.org/onlinedocs/gcc-4.7.4/gfortran/COMPILER_005fVERSION.html) to select the right library.

Assuming, the shared libraries are in a place the compiler can find them, what I had in mind initially was to use a build script to pick the right link flag, depending on the compiler invoked by fpm (passed as an environment variable to the build script).

**Reply by ghost** [reply]
@ivan-pi Hi, I'm also working on a generic _dynload_ fpm package. In case you'd like we can combine efforts and work together 👍.

**Reply by ivan-pi** [reply]
I'd be happy to help. I just noticed @nncarlson already has a dynload package: https://github.com/nncarlson/petaca/blob/master/src/fortran_dynamic_loader.F90
